kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/RecordDecryptor.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption.inband;
+
+import java.nio.ByteBuffer;
+import java.util.function.IntFunction;
+
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.record.Record;
+import org.apache.kafka.common.utils.ByteUtils;
+
+import io.kroxylicious.filter.encryption.AadSpec;
+import io.kroxylicious.filter.encryption.CipherCode;
+import io.kroxylicious.filter.encryption.records.RecordTransform;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+/**
+ * A {@link RecordTransform} that decrypts records that were previously encrypted by {@link RecordEncryptor}.
+ */
+public class RecordDecryptor implements RecordTransform {
+
+    private final IntFunction<DecryptState> keys;
+    int index = 0;
+    private ByteBuffer transformedValue;
+    private Header[] transformedHeaders;
+
+    public RecordDecryptor(IntFunction<DecryptState> keys) {
+        this.keys = keys;
+    }
+
+    @Override
+    public void init(@NonNull Record record) {
+        DecryptState decryptState = keys.apply(index);
```

Comment:
Using a IntFunction to lookup the DecryptState seems fragile  I think the RecordDecryptor would be better just taking a pure function Record -> DecryptState.

Reply:
1. I disagree that it's fragile: the records in a MemoryRecords are perfectly well identified by their index. Think about it: We need two passes anyway, one to resolve the keys and one to do the decryption. So we necessarily require _some_ association between records and `DecryptState`. But it's more efficient to make than an `ArrayList` and use the index, than to force ourselves to use the `Record` as the key and end up with a `HashMap`. 
2. If it helps, we can change the RecordTransformer interface so that it passes this index directly, rather than what we're doing here, which I agree is _messy_.
3. `Record` is an interface which doesn't require it's implementations to override equals/hashCode, so it's a little bit dodgy to rely on it.
4. It so happens that the implementations do override it, but even though they return `duplicated` ByteBuffers for key and value, that doesn't stop those bytes being changed. So `Record` ends up being a mutable key, which I'd argue is not _less_ fragile than using the index.